Python/suggestions.c

```diff
@@ -0,0 +1,230 @@
+    }
+    end = row + len2 - 1;
+    for (i = 0; i < len2 - half; i++) {
+        row[i] = i;
+    }
+
+    /* We don't have to scan two corner triangles (of size len1/2)
+     * in the matrix because no best path can go throught them. This is
+     * not true when len1 == len2 == 2 so the memchr() special case above is
+     * necessary */
+    row[0] = len1 - half - 1;
+    for (i = 1; i < len1; i++) {
+        size_t *p;
+        const char char1 = string1[i - 1];
+        const char *char2p;
+        size_t D, x;
+        /* skip the upper triangle */
+        if (i >= len1 - half) {
+            size_t offset = i - (len1 - half);
+            size_t c3;
+
+            char2p = string2 + offset;
+            p = row + offset;
+            c3 = *(p++) + (char1 != *(char2p++));
+            x = *p;
+            x++;
+            D = x;
+            if (x > c3)
+              x = c3;
+            *(p++) = x;
+        }
+        else {
+            p = row + 1;
+            char2p = string2;
+            D = x = i;
+        }
+        /* skip the lower triangle */
+        if (i <= half + 1) {
+            end = row + len2 + i - half - 2;
+        }
+        /* main */
+        while (p <= end) {
+            size_t c3 = --D + (char1 != *(char2p++));
+            x++;
+            if (x > c3)
+              x = c3;
+            D = *p;
+            D++;
+            if (x > D)
+              x = D;
+            *(p++) = x;
+        }
+        /* lower triangle sentinel */
+        if (i <= half) {
+            size_t c3 = --D + (char1 != *char2p);
+            x++;
+            if (x > c3)
+              x = c3;
+            *p = x;
+        }
+    }
+    i = *end;
+    PyMem_Free(row);
+    return i;
+}
+
+static inline PyObject*
+calculate_suggestions(PyObject* dir,
+                      PyObject* name,
+                      PyObject* oldexceptionvalue)
+{
+    assert(PyList_CheckExact(dir));
+
+    Py_ssize_t dir_size = PyList_GET_SIZE(dir);
+    if (dir_size >= MAX_GETATTR_PREDICT_ITEMS) {
+        return NULL;
+    }
+
+    int suggestion_distance = PyUnicode_GetLength(name);
+    PyObject* suggestion = NULL;
+    for (int i = 0; i < dir_size; ++i) {
+        PyObject *item = PyList_GET_ITEM(dir, i);
+        const char *name_str = PyUnicode_AsUTF8(name);
+        if (name_str == NULL) {
+            PyErr_Clear();
+            continue;
+        }
+        int current_distance = distance(PyUnicode_AsUTF8(name),
+                                        PyUnicode_AsUTF8(item));
+        if (current_distance > MAX_GETATTR_PREDICT_DIST){
+            continue;
+        }
+        if (!suggestion || current_distance < suggestion_distance) {
+            suggestion = item;
+            suggestion_distance = current_distance;
+        }       
+    }
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("%S\n\nDid you mean: %U?",
+                                oldexceptionvalue, suggestion);
```

Comment:
Can you write one line per line, for readability? Something like:

```suggestion
    return PyUnicode_FromFormat("%S\n"
    "\n"
    "Did you mean: %U?",
                                oldexceptionvalue, suggestion);
```